include/utilities/mdarray/iterator.hpp

```diff
@@ -0,0 +1,81 @@
+#ifndef NGEN_MDARRAY_ITERATOR_HPP
+#define NGEN_MDARRAY_ITERATOR_HPP
+
+#include "mdarray.hpp"
+
+namespace ngen {
+
+template<typename T>
+struct mdarray<T>::iterator
+{
+    using iterator_category = std::random_access_iterator_tag;
+    using difference_type   = std::ptrdiff_t;
+    using value_type        = mdarray::value_type;
+    using pointer           = mdarray::pointer;
+    using reference         = mdarray::reference;
+
+    iterator(const mdarray& ref, size_type idx)
+        : m_ref(ref)
+        , m_idx(idx){};
+
+    const_reference operator*() const
```

Comment:
I'm a tad surprised `const_reference` compiles - it's not defined above, so I assume it's pulled in from the enclosing `class mdarray`

Reply:
And I checked that it is used in the test, below

Reply:
It would probably be good to define it explicitly in here, too.